File: 04-Socket-UDP/esempi/02.01-broadcast-js/broadcast-receiver.js
const dgram = require('dgram');

/**
 * Esempio 02.06 - Broadcast UDP Receiver (JavaScript/Node.js)
 * 
 * Questo esempio dimostra come ricevere pacchetti broadcast UDP con Node.js.
 * Il receiver ascolta su una porta specifica e riceve tutti i messaggi
 * broadcast inviati a quella porta sulla rete locale.
 * 
 * Caratteristiche:
 * - Ricezione di messaggi broadcast
 * - Visualizzazione informazioni mittente
 * - Parsing e validazione messaggi
 * 
 * Esecuzione: node broadcast-receiver.js [porta]
 * Esempio: node broadcast-receiver.js 5000
 */

// ==================== CONFIGURAZIONE ====================

const args = process.argv.slice(2);
const DEFAULT_PORT = 5000;

const port = parseInt(args[0]) || DEFAULT_PORT;

// ==================== FUNZIONI UTILITY ====================

/**
 * Restituisce l'ora corrente formattata
 */
function getCurrentTime() {
    const now = new Date();
    return now.toLocaleTimeString('it-IT');
}

/**
 * Parsing del messaggio broadcast
 * Formato atteso: BROADCAST|Messaggio #N|Timestamp: T
 */
function parseBroadcastMessage(message) {
    const parts = message.toString().split('|');
    
    if (parts.length >= 3 && parts[0] === 'BROADCAST') {
        return {
            type: 'BROADCAST',
            content: parts[1],
            timestamp: parts[2]
        };
    }
    
    return {
        type: 'UNKNOWN',
        content: message.toString()
    };
}

// ==================== MAIN ====================

console.log('═'.repeat(55));
console.log('📡 BROADCAST UDP RECEIVER (Node.js)');
console.log('═'.repeat(55));
console.log(`🔌 Porta: ${port}`);
console.log('═'.repeat(55));
console.log();

// Crea socket UDP
const socket = dgram.createSocket('udp4');

let messageCount = 0;

socket.on('error', (err) => {
    console.error(`❌ Errore socket: ${err.message}`);
    console.error('   Verifica che la porta non sia già in uso');
    socket.close();
    process.exit(1);
});

socket.on('listening', () => {
    const address = socket.address();
    console.log(`✅ Socket in ascolto su ${address.address}:${address.port}`);
    console.log('👂 In attesa di messaggi broadcast...');
    console.log();
    console.log('Premi Ctrl+C per terminare');
    console.log();
});

socket.on('message', (msg, remote) => {
    messageCount++;
    
    // Parsing del messaggio
    const parsed = parseBroadcastMessage(msg);
    
    // Visualizza informazioni
    console.log('╔' + '═'.repeat(53));
    console.log(`║ 📥 MESSAGGIO #${messageCount} RICEVUTO`);
    console.log('╠' + '═'.repeat(53));
    console.log(`║ ⏰ Ora: ${getCurrentTime()}`);
    console.log(`║ 👤 Mittente: ${remote.address}:${remote.port}`);
    console.log(`║ 📦 Dimensione: ${msg.length} bytes`);
    console.log('╠' + '═'.repeat(53));
    
    if (parsed.type === 'BROADCAST') {
        console.log(`║ 📝 Tipo: BROADCAST`);
        console.log(`║ 💬 Contenuto: ${parsed.content}`);
        console.log(`║ 🕐 ${parsed.timestamp}`);
    } else {
        console.log(`║ 💬 Contenuto: ${parsed.content}`);
    }
    
    console.log('╚' + '═'.repeat(53));
    console.log();
});

// Binding sulla porta
socket.bind(port);

// ==================== GESTIONE CHIUSURA ====================

function cleanup() {
    console.log('\n\n👋 Chiusura receiver...');
    socket.close();
    console.log('👋 Receiver terminato\n');
    process.exit(0);
}

// Gestione segnali di terminazione
process.on('SIGINT', cleanup);
process.on('SIGTERM', cleanup);

File: 04-Socket-UDP/esempi/02.01-broadcast-java/BroadcastReceiver.java
import java.net.*;
import java.io.*;

/**
 * Esempio 02.02 - Broadcast UDP Receiver (Java)
 * 
 * Questo esempio dimostra come ricevere pacchetti broadcast UDP.
 * Il receiver ascolta su una porta specifica e riceve tutti i messaggi
 * broadcast inviati a quella porta sulla rete locale.
 * 
 * Caratteristiche:
 * - Ricezione di messaggi broadcast
 * - Visualizzazione informazioni mittente
 * - Parsing e validazione messaggi
 * 
 * Compilazione: javac BroadcastReceiver.java
 * Esecuzione: java BroadcastReceiver [porta]
 * Esempio: java BroadcastReceiver 5000
 */
public class BroadcastReceiver {
    private static final int DEFAULT_PORT = 5000;
    private static final int BUFFER_SIZE = 1024;
    
    public static void main(String[] args) {
        int port = DEFAULT_PORT;
        
        // Parsing argomenti da linea di comando
        if (args.length > 0) {
            try {
                port = Integer.parseInt(args[0]);
            } catch (NumberFormatException e) {
                System.err.println("❌ Porta non valida. Uso porta di default: " + DEFAULT_PORT);
            }
        }
        
        System.out.println("═══════════════════════════════════════════════════");
        System.out.println("📡 BROADCAST UDP RECEIVER");
        System.out.println("═══════════════════════════════════════════════════");
        System.out.println("🔌 Porta: " + port);
        System.out.println("═══════════════════════════════════════════════════\n");
        
        try (DatagramSocket socket = new DatagramSocket(port)) {
            System.out.println("✅ Socket in ascolto sulla porta " + port);
            System.out.println("👂 In attesa di messaggi broadcast...\n");
            System.out.println("Premi Ctrl+C per terminare\n");
            
            byte[] buffer = new byte[BUFFER_SIZE];
            int messageCount = 0;
            
            while (true) {
                // Crea il pacchetto per la ricezione
                DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                
                // Riceve il pacchetto (bloccante)
                socket.receive(packet);
                messageCount++;
                
                // Estrae i dati dal pacchetto
                String message = new String(
                    packet.getData(),
                    0,
                    packet.getLength()
                );
                
                InetAddress senderAddress = packet.getAddress();
                int senderPort = packet.getPort();
                
                // Visualizza informazioni sul messaggio ricevuto
                System.out.println("╔═══════════════════════════════════════════════════");
                System.out.printf("║ 📥 MESSAGGIO #%d RICEVUTO%n", messageCount);
                System.out.println("╠═══════════════════════════════════════════════════");
                System.out.printf("║ ⏰ Ora: %s%n", getCurrentTime());
                System.out.printf("║ 👤 Mittente: %s:%d%n", senderAddress.getHostAddress(), senderPort);
                System.out.printf("║ 📦 Dimensione: %d bytes%n", packet.getLength());
                System.out.println("╠═══════════════════════════════════════════════════");
                
                // Parsing del messaggio
                parseBroadcastMessage(message);
                
                System.out.println("╚═══════════════════════════════════════════════════\n");
            }
            
        } catch (SocketException e) {
            System.err.println("❌ Errore creazione socket: " + e.getMessage());
            System.err.println("   Verifica che la porta non sia già in uso");
        } catch (IOException e) {
            System.err.println("❌ Errore ricezione pacchetto: " + e.getMessage());
        }
        
        System.out.println("\n👋 Receiver terminato");
    }
    
    /**
     * Parsing del messaggio broadcast
     * Formato atteso: BROADCAST|Messaggio #N|Timestamp: T
     */
    private static void parseBroadcastMessage(String message) {
        String[] parts = message.split("\\|");
        
        if (parts.length >= 3 && parts[0].equals("BROADCAST")) {
            System.out.println("║ 📝 Tipo: BROADCAST");
            System.out.println("║ 💬 Contenuto: " + parts[1]);
            System.out.println("║ 🕐 " + parts[2]);
        } else {
            System.out.println("║ 💬 Contenuto: " + message);
        }
    }
    
    /**
     * Restituisce l'ora corrente formattata
     */
    private static String getCurrentTime() {
        java.time.LocalTime now = java.time.LocalTime.now();
        return String.format("%02d:%02d:%02d",
            now.getHour(),
            now.getMinute(),
            now.getSecond()
        );
    }
}
